src/Application/Commands/RedistributeMatchFunds.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MatchBot\Application\Commands;
+
+use MatchBot\Domain\CampaignFundingRepository;
+use MatchBot\Domain\DonationRepository;
+use MatchBot\Domain\FundingWithdrawal;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Redistribute match funding allocations where possible, from lower to higher priority match fund pots.
+ */
+class RedistributeMatchFunds extends LockingCommand
+{
+    protected static $defaultName = 'matchbot:redistribute-match-funds';
+
+    public function __construct(
+        private CampaignFundingRepository $campaignFundingRepository,
+        private DonationRepository $donationRepository
+    ) {
+        parent::__construct();
+    }
+
+    protected function configure(): void
+    {
+        $this->setDescription('Moves match funding allocations from lower to higher priority funds where possible');
+    }
+
+    protected function doExecute(InputInterface $input, OutputInterface $output): int
+    {
+        // TODO Change the fixed lookback to 2 days, or parameter-ise it, once CC23 is tidied in Prod.
+        $donationsToCheck = $this->donationRepository->findWithMatchingWhichCouldBeReplacedWithHigherPriorityAllocation(
+            new \DateTimeImmutable('-3 weeks')
+        );
+
+        $donationsAmended = 0;
+        foreach ($donationsToCheck as $donation) {
+            $highestAllocationOrderUsedForDonation = 0;
+            foreach ($donation->getFundingWithdrawals() as $withdrawal) {
+                $funding = $withdrawal->getCampaignFunding();
+                if (!$funding) {
+                    // Not entirely sure why this is nullable at all; we can't work on withdrawals without a funding.
+                    continue;
+                }
+
+                $highestAllocationOrderUsedForDonation = max(
+                    $highestAllocationOrderUsedForDonation,
+                    $funding->getAllocationOrder(),
+                );
+            }
+
+            $fundings = $this->campaignFundingRepository->getAvailableFundings($donation->getCampaign());
+
+            $fundingsAllowForRedistribution = false;
+            foreach ($fundings as $funding) {
+                if ($funding->getAllocationOrder() <= $highestAllocationOrderUsedForDonation) {
+                    continue;
+                }
+
+                $amountToAllocate = min($donation->getAmount(), $funding->getAmountAvailable());
+                if ($amountToAllocate <= 0) {
+                    continue;
+                }
+
+                $fundingsAllowForRedistribution = true;
+                break;
+            }
+
+            if (!$fundingsAllowForRedistribution) {
+                continue;
+            }
+
+            $this->donationRepository->releaseMatchFunds($donation);
+            $this->donationRepository->allocateMatchFunds($donation);
```

Comment:
Two questions about this:

1. Is it possible for match funds to be used up by another thread in between these two lines? What happens in that case? Might we need to have a lock that covers both or something to prevent that, or if not do we know how we'll deal with the issue when it happens.

2. Could we combine it into one function call to the donationRepo class? I think that might be clearer, and it can do the two things internally - maybe:

```suggestion
            $this->donationRepository->releaseAndReallocateMatchFunds($donation);
```

Reply:
1. I think unfortunately yes. We have nothing that locks allocation across funds in this way and I think it could be very disruptive, so there is a trade-off here. I've now added an error log to tell us if it happens. As already planned we will run this at quiet times – we might also want to talk again in the team about the risk vs. benefit before deciding to apply to it to still-open campaigns, since the only advantage is data charities see (which we could always caveat or fudge another way).
2. I don't think this gains us any lock safety, so my preference is to keep the separate-ness and slight risk clear for now.

Reply:
Adding the error log sounds good. If we start seeing the error in practice we can think about using a lock - I think it should be fairly straightforward to lock creation of new donations during the time this is running, probably using either https://dev.mysql.com/doc/refman/8.0/en/locking-functions.html#function_get-lock or https://symfony.com/doc/current/components/lock.html#available-stores .